[16.0][IMP] sale_order_lot_selection: add lot selection wizard

halbtonjazz opened the pull request:
Implement lot selection wizard for sales quotations to enable creating quotations with specific lot/serial numbers directly from stock lots. This improves traceability and allows precise lot selection for industries requiring specific serial number tracking.

Task: 4470

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced a new wizard that streamlines adding stock lots to sale orders, allowing users to attach lots to an existing order or create a new one.
  - Added a configuration setting to enable the generation of sale orders from lots.
  - Included a new "Lots/Serial Numbers" menu option for easier management of production lots.

- **Enhancements**
  - Updated module configurations and improved security handling to support the extended functionalities.
  - Refined UI elements in module descriptions for greater clarity.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

halbtonjazz committed:
[IMP] sale_order_lot_selection: add lot selection wizard

Implements a wizard interface for managing lot selection in sale orders:
- Add configuration option to enable/disable lot selection from stock
- Introduce wizard to select multiple lots with quantities
- Add security rules and access rights for lot operations
- Enhance UX with dynamic quantity validation against available stock

Task: 4470

CetmixGitDrone commented:
Hi @bodedra,
some modules you are maintaining are being modified, check this out!

halbtonjazz committed:
[IMP] sale_order_lot_selection: add lot selection wizard

- Refactor the wizard to handle new instances, improving reliability.
- Add validations to ensure a sale order or customer is selected, preventing errors.
- Update the UI to reflect new functionality and provide better user guidance.
- Set default lot quantities based on product quantities for accuracy.

Task: 4470

halbtonjazz committed:
[ADD] sale_order_lot_selection: add tests

Add test cases to verify functionality of creating sale orders from stock
lots and adding lots to existing orders

Task: 4470

halbtonjazz committed:
[IMP] sale_order_lot_selection: update docs

Add copyright headers to Python and XML files.
Update usage instructions to reflect the new lot selection wizard
feature and configuration options.
Add Cetmix contributors and new maintainer to module metadata.

Task: 4470

halbtonjazz committed:
[ADD] sale_order_lot_selection: add use cases

Add business use cases and context to the documentation to explain
why selecting specific lot or serial numbers when creating
quotations is needed.
Describes scenarios where product traceability, compliance, or
customer-specific requirements are essential.

Task: 4470

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+            for line_id in self.line_ids
+        ]
+
+    def _action_open_sale_order(self):
+        """Open the sale order form view."""
+        return {
+            "type": "ir.actions.act_window",
+            "name": "Sale Order",
+            "res_model": "sale.order",
+            "view_mode": "form",
+            "res_id": self.sale_order_id.id,
+            "target": "current",
+        }
+
```
Type hints are not standard in Odoo/OCA modules

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        )
+        return self._action_open_sale_order()
+
+    def _get_vals_for_add_lot_in_sale_order_line(self):
+        """Get values for adding lot in sale order line."""
+        if not self.line_ids:
+            raise ValidationError(_("Please select at least one lot"))
+        return [
+            (
+                0,
+                0,
+                {
+                    "lot_id": line_id.lot_id.id,
+                    "product_id": line_id.lot_id.product_id.id,
+                    "product_uom_qty": line_id.quantity,
+                },
+            )
+            for line_id in self.line_ids
+        ]
+
```
Type hints are not standard in Odoo/OCA modules

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        )
+        return self._action_open_sale_order()
+
+    def action_create_sale_order(self):
+        """Create a sale order from the selected lots."""
+        self.ensure_one()
+        if not self.partner_id:
+            raise ValidationError(_("Please select a Customer"))
+        self.sale_order_id = self.env["sale.order"].create(
+            {
+                "partner_id": self.partner_id.id,
+                "order_line": self._get_vals_for_add_lot_in_sale_order_line(),
+            }
+        )
+        return self._action_open_sale_order()
+
```
Type hints are not standard in Odoo/OCA modules, and the trailing comma is already present in the code

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        compute="_compute_partner_id",
+    )
+
+    @api.depends("sale_order_id")
+    def _compute_partner_id(self):
+        for record in self:
+            record.partner_id = record.sale_order_id.partner_id.id
+
```
Type hints are not standard in Odoo/OCA modules

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+    max_qty = fields.Float(string="Max Quantity", related="lot_id.product_qty")
+
+    @api.constrains("quantity")
+    def _check_quantity(self):
+        """Check that the quantity is not greater than the max quantity."""
+        for record in self:
+            if record.quantity > record.max_qty:
+                raise ValidationError(
+                    _(
+                        "The quantity of lot %(lot_name)s cannot be greater than "
+                        "the available quantity %(max_qty)s.",
+                        lot_name=record.lot_id.name,
+                        max_qty=record.max_qty,
+                    )
+                )
+
```
Type hints are not standard in Odoo/OCA modules, and all required trailing commas are already present in the code.

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+            "target": "new",
+        }
+
+    def action_add_lots_to_sale_order(self):
+        """Add selected lots to the existing sale order."""
+        self.ensure_one()
+        if not self.sale_order_id:
+            raise ValidationError(_("Please select a Sale Order"))
+        self.sale_order_id.write(
+            {
+                "order_line": self._get_vals_for_add_lot_in_sale_order_line(),
+            }
+        )
+        return self._action_open_sale_order()
+
```
Type hints are not standard in Odoo/OCA modules, and all required trailing commas are already present in the code.

halbtonjazz commented on sale_order_lot_selection/models/res_config_settings.py:
```diff
@@ -0,0 +1,14 @@
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
```
Thanks, I'll fix that in a new commit

halbtonjazz commented on sale_order_lot_selection/view/res_config_settings.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="res_config_settings_view_form_stock" model="ir.ui.view">
+        <field
+            name="name"
+        >res.config.settings.view.form.inherit.product.expiry.stock</field>
+        <field name="model">res.config.settings</field>
+        <field name="inherit_id" ref="stock.res_config_settings_view_form" />
+        <field name="arch" type="xml">
```
Thanks, I'll fix that in a new commit

halbtonjazz committed:
[REF] sale_order_lot_selection: refactor wizard

Remove maintainer key as modules are not actively maintained.
Update view name in res_config_settings to better reflect module purpose.
Apply code review suggestions from coderabbitai bot.

Task: 4470
